design/garbage-collection.md

```diff
@@ -39,18 +77,85 @@ In the initial state, both clients have `"ab"`.
 
 #### State 4
 
 ![garbage-collection-4](media/garbage-collection-4.png)
 
-Finally, after client `c1` receives change `4` from server, purges textnode `"b"` because it is no longer referenced remotely.
+`Client a` pushpull but nothing to push or pull. `minVersionVector` is still `{a:1, b:1}`, so no GC happens.
+
+#### State 5
+
+![garbage-collection-4](media/garbage-collection-5.png)
+
+`Client b` pushpull but nothing to push or pull. `minVersionVector` is now `{a:3, b:1}`, node "b" removedAt is `3@a`, and `minVersionVector[a] = 3 >= 3`, thus `client b` meets GC condition 
+
+#### State 6
+
+![garbage-collection-4](media/garbage-collection-6.png)
+
+`Client a` pushpull but nothing to push or pull. `minVersionVector` is now `{a:3, b:1}`, node "b" removedAt is `3@a`, and `minVersionVector[a] = 3 >= 3`, thus `client a` meets GC condition
+
+### How we handle if min version vector includes detached client's lamport.
+We have to wipe out detached client's lamport from every version vector in db and other client's local version vector.
+![detached-user-version-vector](media/detached-user-version-vector.jpg)
+
+For example,
+```
+// initial state
+c1.versionVector = {c1:4, c2:3, c3: 4}
+c2.versionVector = {c1:3, c2:4, c3: 5}
+c3.versionVector = {c1:3, c2:3, c3: 6}
+
+db.versionVector = {
+    c1: {c1:4, c2:3, c3: 4},
+    c2: {c1:3, c2:4, c3: 5},
+    c3: {c1:3, c2:3, c3: 6}
+}
+
+// process
+1. c1 detach and remove its version vector from db.
+
+db.versionVector = {
+    c2: {c1:3, c2:4, c3: 5}
+    c3: {c1:3, c2:3, c3: 6}
+}
+
+2. compute minVersionVector
+min(c2.vv, c3.vv) = min({c1:3, c2:4, c3: 5}, {c1:3, c2:3, c3:5}) = {c1:3, c2:3, c3:5}
+
+```
+as you can see from above, c1's lamport is still inside minVersionVector, and also in every client's local document's version vector too.
+
+So we need to filter detached client's lamport from
+1. db.version vector
+2. other client's local version vector.
+
+But it causes n+1 query problem to remove lamport from db.versionVector. So we choose remove only client's version vector from table, and filter minVersionVector by active clients.
+
+```
+// initial state
+db.versionVector = {
+    c2: {c1:3, c2:4, c3: 5},
+    c3: {c1:3, c2:3, c3: 6}
+}
+min(c2.vv, c3.vv) = min({c1:3, c2:4, c3: 5}, {c1:3, c2:3, c3:5}) = 
+{c1:3, c2:3, c3:5}
+c2, c3 are acitve(attached).
+
+minVersionVector = {c1:3, c2:3, c3:5}.Filter([c1]) = {c2:3, c3:5}
+```
+
+After client receive this minVersionVector, it will filter its version vector to remove detached client's lamport.
+The next pushpull request will contains filtered version vector so that eventually db.version vector will store attached client's version vector only.
+![filter-version-vector](media/filter-version-vector.jpg)
```

Comment:
I think the image is different from the code above. If the image sync with the code above, It would be easier to understand!

Reply:
This code filtered `c1`'s vv. i think this code will be filterd `c3`'s vv
```
minVersionVector = {c1:3, c2:3, c3:5}.Filter([c1]) = {c1:3, c2:3}
```